File: executor/common/src/cond_stmt.rs
use super::{cond_stmt_base::CondStmtBase, tag::TagSeg,cond_state::CondState};
use std::hash::{Hash, Hasher};
use serde_derive::{Deserialize, Serialize};

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct CondStmt {
    pub base: CondStmtBase,
    pub offsets: Vec<TagSeg>,
    pub offsets_opt: Vec<TagSeg>,
    pub variables: Vec<u8>,

    pub speed: u32,
    pub is_desirable: bool, // non-convex
    pub is_consistent: bool,
    pub fuzz_times: usize,
    pub state: CondState,
    pub num_minimal_optima: usize,
    pub linear: bool,
}

impl PartialEq for CondStmt {
    fn eq(&self, other: &CondStmt) -> bool {
        self.base == other.base
    }
}

impl Eq for CondStmt {}

impl Hash for CondStmt {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.base.cmpid.hash(state);
        self.base.context.hash(state);
        self.base.order.hash(state);
    }
}

impl CondStmt {
    pub fn new() -> Self {
        let cond_base = Default::default();
        Self {
            base: cond_base,
            offsets: vec![],
            offsets_opt: vec![],
            variables: vec![],
            speed: 0,
            is_consistent: true,
            is_desirable: true,
            fuzz_times: 0,
            state: CondState::default(),
            num_minimal_optima: 0,
            linear: false,
        }
    }

    pub fn from(cond_base: CondStmtBase) -> Self {
        let mut cond = Self::new();
        cond.base = cond_base;
        cond
    }
}

File: fuzz_checker/cond_stmt.py
from cond_stmt_base import CondStmtBase
import defs

class CondStmt:
    def __init__(self):
        self.offsets = self.offsets_opt = self.variables = self.speed = self.is_desirable = self.is_consistent = self.fuzz_times = self.state = self.num_minimal_optima = self.linear= 0
        self.base = CondStmtBase
        self.skipping = False


    @staticmethod
    def fromJson(json):
        condStmt = CondStmt()
        condStmt.__dict__.update(json)
        condStmt.base = CondStmtBase.fromJson(json['base'])
        return condStmt

    def is_one_byte(self) -> bool:
        if len(self.offsets) == 1 and self.offsets[0]['end'] - self.offsets[0]['begin'] == 1:
            return True
        return False

    def isSkipped(self):
        return self.skipping or ((self.base.op & defs.COND_BASIC_MASK) == defs.COND_FN_OP)